elements/lisk-dpos/test/unit/vote_weight_snapshot.spec.ts

```diff
@@ -0,0 +1,870 @@
+				.fn()
+				.mockReturnValue({ totalEarning: BigInt(0), totalBurnt: BigInt(0) }),
+			dataAccess: {
+				getBlockHeadersByHeightBetween: jest.fn().mockResolvedValue([]),
+				getConsensusState: jest.fn().mockResolvedValue(undefined),
+				getDelegateAccounts: jest.fn().mockResolvedValue([]),
+				getDelegates: jest.fn().mockResolvedValue([]),
+			},
+		};
+		dpos = new Dpos({ chain: chainStub });
+		stateStore = new StateStoreMock([], {});
+	});
+
+	describe('given genesis block', () => {
+		let delegates: Account[];
+		let genesisBlock: Block;
+
+		beforeEach(async () => {
+			// Arrange
+			delegates = getDelegateAccounts(103);
+			for (const delegate of delegates) {
+				delegate.totalVotesReceived = BigInt(10) ** BigInt(12);
+			}
+			genesisBlock = {
+				id: 'genesis-block',
+				timestamp: 10,
+				height: 1,
+				generatorPublicKey: forgers[0].publicKey,
+				reward: BigInt(500000000),
+				totalFee: BigInt(100000000),
+				transactions: [],
+			} as Block;
+			stateStore = new StateStoreMock([...delegates]);
+		});
+
+		describe('when apply is called', () => {
+			it('should snapshot for first round and offset rounds', async () => {
+				// Act
+				await dpos.apply(genesisBlock, stateStore);
+
+				// Assert
+				const voteWeightsStr = await stateStore.consensus.get(
+					CONSENSUS_STATE_VOTE_WEIGHTS_KEY,
+				);
+				const voteWeights = JSON.parse(voteWeightsStr as string);
+				expect(voteWeights).toHaveLength(3);
+				expect(voteWeights[0].round).toEqual(1);
+				expect(voteWeights[1].round).toEqual(2);
+				expect(voteWeights[2].round).toEqual(3);
+			});
+		});
+	});
+
+	describe('given not the last block of a round', () => {
+		let delegates: Account[];
+		let genesisBlock: Block;
+
+		beforeEach(async () => {
+			// Arrange
+			delegates = getDelegateAccounts(103);
+			for (const delegate of delegates) {
+				delegate.totalVotesReceived = BigInt(10) ** BigInt(12);
+			}
+			genesisBlock = {
+				id: 'random-block',
+				timestamp: 50,
+				height: 5,
+				generatorPublicKey: forgers[0].publicKey,
+				reward: BigInt(500000000),
+				totalFee: BigInt(100000000),
+				transactions: [],
+			} as Block;
+			stateStore = new StateStoreMock([forgers[0], ...delegates]);
+			jest.spyOn(stateStore.consensus, 'set');
+		});
+
+		describe('when apply is called', () => {
+			it('should not snapshot the voteweight', async () => {
+				// Act
+				await dpos.apply(genesisBlock, stateStore);
+
+				// Assert
+				expect(stateStore.consensus.set).not.toHaveBeenCalled();
+				expect(chainStub.dataAccess.getDelegates).not.toHaveBeenCalled();
+			});
+		});
+	});
+
+	describe('given the last block of a round', () => {
+		let delegates: Account[];
+		let block: Block;
+
+		describe('when there are changes in the last block', () => {
+			let updatedDelegate: Account;
+
+			beforeEach(async () => {
+				delegates = getDelegateAccounts(200);
+				for (const delegate of delegates) {
+					delegate.totalVotesReceived = randomBigInt(900, 5000);
```

Comment:
hmmmm, so this is the self-votes?

Reply:
ok, I see that there is actually the multiplication by 10^9 in the `randomBigInt` function. (I personally find this a bit confusing)

Reply:
maybe ill rename this function to `randomBigIntWithPowerof8`?

Reply:
if you want. Or have the function return a random integer and multiply by 10^8 explicitly in this line.
I don't have a strong opinion on this.